compiler/z/codegen/FPTreeEvaluator.cpp

```diff
@@ -1121,45 +1116,36 @@ OMR::Z::TreeEvaluator::floatRemHelper(TR::Node * node, TR::CodeGenerator * cg)
+      TR::Node::recreate(node, TR::dcall);
       }
-   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, labelNotExact);
-
-   void * remHelper ;
-   if (node->getDataType() == TR::Float)
-     remHelper = (void *) cg->symRefTab()->findOrCreateRuntimeHelper(TR_S390floatRemainder, false, false, false)->getMethodAddress();
-   else
-     remHelper = (void *) cg->symRefTab()->findOrCreateRuntimeHelper(TR_S390doubleRemainder, false, false, false)->getMethodAddress();
-
-   TR::RegisterDependencyConditions * postDeps = generateRegisterDependencyConditions(0, 12, cg);
 
-   genLoadAddressConstant(cg, node, (uintptrj_t) remHelper, raReg, NULL, postDeps, NULL);
-
-   cursor = generateRRInstruction(cg, TR::InstOpCode::BASR, node, raReg, raReg);
-
-   TR::Linkage * linkage = cg->getS390Linkage();
-   postDeps->addPostCondition(raReg, linkage->getReturnAddressRegister());
-   postDeps->addPostCondition(targetRegister, TR::RealRegister::FPR0);
-   postDeps->addPostCondition(firstRegister, TR::RealRegister::FPR1);
-   postDeps->addPostCondition(secondRegister, TR::RealRegister::FPR2);
-   postDeps->addPostCondition(dummyRegister3, TR::RealRegister::FPR3);
-   postDeps->addPostCondition(dummyRegister4, TR::RealRegister::FPR4);
-   postDeps->addPostCondition(dummyRegister5, TR::RealRegister::FPR5);
-   postDeps->addPostCondition(dummyRegister6, TR::RealRegister::FPR6);
-   postDeps->addPostCondition(dummyRegister7, TR::RealRegister::FPR7);
+   // Putting call-out part in the OOL as we only need to call helper when result is not exact which is very rare.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(labelNotExact,labelOK,cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, labelNotExact);
+   TR::Linkage *linkage = cg->createLinkage(TR_System);
+   node->setSymbolReference(cg->symRefTab()->findOrCreateRuntimeHelper(node->getOpCodeValue() == TR::fcall ? TR_S390jitMathHelperFREM : TR_S390jitMathHelperDREM, false, false, false));
+   TR::Register *helperReturnRegister = linkage->buildSystemLinkageDispatch(node);
+   // In addition to returning the result register, above dispatch call also set node with result register.
+   // Main line sets result in targetRegister so in OOL we copy helperReturnRegister to targetRegistert and unset node register.
+   generateRRInstruction(cg, node->getDataType() == TR::Float ? TR::InstOpCode::LER : TR::InstOpCode::LDR, node,  targetRegister, helperReturnRegister);
+   node->unsetRegister();
+   cg->stopUsingRegister(helperReturnRegister);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, labelOK);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+
+   TR::RegisterDependencyConditions * postDeps = generateRegisterDependencyConditions(0, 4, cg);   
```

Comment:
Can we mark the ICF in which this post dependency is being used with `setStartInternalControlFlow` / `setEndInternalControlFlow`?

Reply:
We are using OOL path to generate a call out sequences so there is no ICF in OOL and in mainline we postDeps are attached to `labelOK` which is done Label for this node. This seemed to me like a simple case of ICF and I thought attaching postDeps to `labelOK` is sufficient.  Do we still need to mark instructions as setStart/EndInternalControlFlow ?